### PR TITLE
fix(common): do not replace zero/empty string with null request body …

### LINKED ISSUE #19195: HttpClient posting number zero not working
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

Number zero (0) doesn't get sent via post - body.
All other numbers seem to work fine.

## Expected behavior

It should be possible to post a 0 via post - body.

## Minimal reproduction of the problem with instructions

DEMO: https://plnkr.co/edit/YYQfVJYAJjWyy6eMaR5Q?p=preview

Open dev tools and look at the request. 

## Environment

<pre><code>
Angular version: 4.3.6

Browser:
- [x] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX

</code></pre>

### PR DESCRIPTION
**Pull request by snerks**
…in HttpClient.post

'body' parameter to HttpClient.post method was replaced by null if it evaluated as falsy

Fixes #19195

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number:  #19195

## What is the new behavior?
The values `0` and `''` (empty string) are not replaced by `null` as the request body in the `HttpClient.post` method.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/common/http/src/client.ts [hunk 1]
```diff
@@ -371,7 +371,9 @@ export class HttpClient {
       }
 
       // Construct the request.
-      req = new HttpRequest(first, url !, options.body || null, {
```
Comment (vicb) [inline packages/common/http/src/client.ts]: options.body === undefined ? null : options.body

**Comment by vicb** [comment]
Thanks for your PR, a few comments:
- the code is not optimal (see comment),
- you seem to add a feature in a fix commit (echoRequest) should be in a different commit/PR
- you need to run `gulp public-api:update` when you change the public API - which explains why the CI fails.

File: packages/common/http/src/client.ts (hunk 1 above)
Comment (snerks) [reply packages/common/http/src/client.ts]: I think the suggested test would not filter `false` or `NaN`.

The intent here is to behave as before but allow `0` and `''` (empty string) through the filter.

**Comment by snerks** [comment]
@vicb - thanks for reviewing.

I didn't realise that the `testing` `request.ts` file (packages/common/http/testing/src/request.ts) was part of the public API? Can you confirm?

If so, the intent was to allow tests to signal that the request body should be returned as the response body.
Is there a way to do that without affecting the Public API?

**Comment by vicb** [comment]
@alxhub will review in more detail.

you need to sign the CLA with the address used for the commits (martin.github@...)

**Comment by mayrmax** [comment]
any updates on this? this needed for a important feature on our production application.

**Comment by snerks** [comment]
I signed it!

**Comment by slawojstanislawski** [comment]
It seems there's been no activity on it for some time, can it please be merged?

**Comment by trotyl** [comment]
Likely replaced by https://github.com/angular/angular/pull/19958.

**Comment by mhevery** [comment]
#19958 fixes this issue.

**Pull request closed without merging**